File: src/main/ruby/diffj/ast/package.rb
#!/usr/bin/jruby -w
# -*- ruby -*-

require 'rubygems'
require 'riel'
require 'java'
require 'diffj/ast/element'

include Java

module DiffJ
  class PackageComparator < ElementComparator
    include Loggable
    
    PACKAGE_REMOVED = "package removed: {0}"
    PACKAGE_ADDED = "package added: {0}"
    PACKAGE_RENAMED = "package renamed from {0} to {1}"

    def find_name_node parent
      parent.find_child "net.sourceforge.pmd.ast.ASTName"
    end

    def find_first_child parent
      parent.find_child
    end

    def compare_names fromnode, tonode
      fromname = find_name_node fromnode
      toname = find_name_node tonode

      if fromname.to_string != toname.to_string
        changed fromname, toname, PACKAGE_RENAMED
      end
    end

    def compare fromcu, tocu
      frompkg = fromcu.package
      topkg = tocu.package

      if frompkg
        if topkg
          compare_names frompkg, topkg
        else
          fromname = find_name_node frompkg
          topos = find_first_child(tocu) || tocu
          deleted fromname, topos, PACKAGE_REMOVED
        end
      elsif topkg
        toname = find_name_node topkg
        frompos = find_first_child(fromcu) || fromcu
        added frompos, toname, PACKAGE_ADDED
      end
    end
  end
end

File: src/main/java/org/incava/diffj/compunit/Package.java
package org.incava.diffj.compunit;

import net.sourceforge.pmd.ast.ASTCompilationUnit;
import net.sourceforge.pmd.ast.ASTName;
import net.sourceforge.pmd.ast.ASTPackageDeclaration;
import net.sourceforge.pmd.ast.SimpleNode;
import org.incava.diffj.element.Differences;
import org.incava.ijdk.text.Message;
import org.incava.pmdx.CompilationUnitUtil;
import org.incava.pmdx.SimpleNodeUtil;

public class Package {
    public static final Message PACKAGE_REMOVED = new Message("package removed: {0}");
    public static final Message PACKAGE_ADDED = new Message("package added: {0}");
    public static final Message PACKAGE_RENAMED = new Message("package renamed from {0} to {1}");

    private final ASTCompilationUnit compUnit;
    private final ASTPackageDeclaration pkg;

    public Package(ASTCompilationUnit compUnit) {
        this.compUnit = compUnit;
        this.pkg = CompilationUnitUtil.getPackage(compUnit);
    }

    public boolean isEmpty() {
        return pkg == null;
    }

    public void diff(Package toPackage, Differences differences) {
        if (isEmpty()) {
            if (!toPackage.isEmpty()) {
                ASTName    name    = toPackage.getPackageName();
                SimpleNode fromPos = getChild();
                differences.added(fromPos, name, PACKAGE_ADDED);
            }
        }
        else if (toPackage.isEmpty()) {
            ASTName    name  = getPackageName();
            SimpleNode toPos = toPackage.getChild();
            differences.deleted(name, toPos, PACKAGE_REMOVED);
        }
        else {
            ASTName fromName = getPackageName();
            String  fromStr  = SimpleNodeUtil.toString(fromName);
            ASTName toName   = toPackage.getPackageName();
            String  toStr    = SimpleNodeUtil.toString(toName);

            if (!fromStr.equals(toStr)) {
                differences.changed(fromName, toName, PACKAGE_RENAMED);
            }
        }
    }

    protected SimpleNode getChild() {
        SimpleNode child = SimpleNodeUtil.findChild(compUnit);
        return child == null ? compUnit : child;
    }

    public ASTName getPackageName() {
        return SimpleNodeUtil.findChild(pkg, net.sourceforge.pmd.ast.ASTName.class);
    }
}
